### PR TITLE
[target-allocator] use json iterator to marshal in server

### PR DESCRIPTION
**Pull request by kristinapathak**
fixes https://github.com/open-telemetry/opentelemetry-operator/issues/1336

This time, server functionality is covered by unit tests.  Below are the results of the benchmarks, showing using the standard json library (std) vs jsoniter.

[test_results.txt](https://github.com/open-telemetry/opentelemetry-operator/files/10541897/test_results.txt)

### CODE REVIEW

File: cmd/otel-allocator/server/server.go [hunk 1]
```diff
@@ -174,6 +185,14 @@ func (s *Server) errorHandler(w http.ResponseWriter, err error) {
 }
 
 func (s *Server) jsonHandler(w http.ResponseWriter, data interface{}) {
+	w.Header().Set("Content-Type", "application/json")
+	err := s.jsonMarshaller.NewEncoder(w).Encode(data)
+	if err != nil {
```
Comment (jaronoff97) [inline cmd/otel-allocator/server/server.go:190]: do our tests already cover this case? i.e. is this change just going to work with what we do in the collector? using specifically the [yaml unmarshal](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/receiver/prometheusreceiver/metrics_receiver.go#L224) function?

**Review by jaronoff97 — COMMENTED** [review]
after this change goes out, should we upstream a change to the collector?

File: cmd/otel-allocator/server/server.go (hunk 1 above)
Comment (kristinapathak) [reply cmd/otel-allocator/server/server.go:190]: That yaml unmarshal function is for a response from the scrape config handler, which doesn't use the `jsonHandler` function.  [Here](https://github.com/open-telemetry/opentelemetry-operator/blob/6ca223ad8354d5f37b0c29489a155de751cdd2ef/cmd/otel-allocator/server/server.go#L100) is the marshaling logic for the scrape config endpoint.

The `jsonHandler()` function is used for the job and targets handlers, both of which have test cases which ensure that this marshaling is still compatible with `json.Unmarshal()`:
- https://github.com/open-telemetry/opentelemetry-operator/blob/6ca223ad8354d5f37b0c29489a155de751cdd2ef/cmd/otel-allocator/server/server_test.go#L175
- https://github.com/open-telemetry/opentelemetry-operator/blob/6ca223ad8354d5f37b0c29489a155de751cdd2ef/cmd/otel-allocator/server/server_test.go#L532

**Comment by kristinapathak** [comment]
Latest benchmarks
[Uploading test_results.txt…]()

**Review by jaronoff97 — APPROVED** [review]

**Pull request merged**